File: ext/seo-fixer.js
import { Glob } from "bun";
import { load } from 'cheerio';
import path from 'path';
import sharp from 'sharp';

// --- HELPER UTILS ---
const escapeHtmlAttr = (text) => {
  if (!text) return '';
  return text
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '’')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
};

const prepareDesc = (text) => escapeHtmlAttr(text.replace(/\s+/g, ' ').trim());

// --- CORE FUNCTIONS ---

async function mirrorAndConvert(externalUrl, baseUrl) {
  try {
    const url = new URL(externalUrl);
    const baseHostname = new URL(baseUrl).hostname;

    // Filter Internal atau URL Skema
    if (url.hostname === baseHostname || url.hostname === 'localhost' || url.hostname === 'schema.org') {
      return externalUrl.replace(baseUrl, '');
    }

    const ext = path.extname(url.pathname).toLowerCase();
    const isSvg = ext === '.svg';
    const finalExt = isSvg ? '.svg' : '.webp';

    // Path Management Lokal
    const localPathName = ext ? url.pathname.replace(ext, finalExt) : `${url.pathname}${finalExt}`;
    const localPath = path.join('img', url.hostname, localPathName);

    // Bun.file check jauh lebih cepat
    const fileTarget = Bun.file(localPath);
    if (await fileTarget.exists()) return `/${localPath.replace(/\\/g, '/')}`;

    // Download pakai Fetch Native Bun (Ganti Axios)
    const response = await fetch(externalUrl, {
      headers: { 'User-Agent': 'Mozilla/5.0' }
    });
    if (!response.ok) throw new Error('Download Gagal');

    const arrayBuffer = await response.arrayBuffer();
    const buffer = Buffer.from(arrayBuffer);

    if (isSvg) {
      await Bun.write(localPath, buffer);
    } else {
      // Pastikan folder ada (Bun.write otomatis bikin folder, tapi Sharp butuh foldernya ada)
      const dirPath = path.dirname(localPath);
      await $`mkdir -p ${dirPath}`; // Gunakan Shell Bun buat kilat
      await sharp(buffer).webp({ quality: 85 }).toFile(localPath);
    }

    return `/${localPath.replace(/\\/g, '/')}`;
  } catch (err) {
    return externalUrl;
  }
}

async function processFile(file, baseUrl) {
  const rawContent = await Bun.file(file).text();
  const baseName = path.basename(file);
  console.log(`🚀 SEO Turbo: ${baseName}`);

  const $ = load(rawContent, { decodeEntities: false });
  const head = $('head');

  // 1. MIRRORING GAMBAR DALAM BODY (Paralel di dalam file)
  const imgPromises = $('img').toArray().map(async (el) => {
    const src = $(el).attr('src');
    if (src && src.startsWith('http')) {
      const localPath = await mirrorAndConvert(src, baseUrl);
      if (localPath.startsWith('/')) $(el).attr('src', `${baseUrl}${localPath}`);
    }
  });
  await Promise.all(imgPromises);

  // 2. LOGIKA DATA SEO
  const articleTitle = $('title').text().split(' - ')[0].trim() || 'Layar Kosong';
  const escapedTitle = escapeHtmlAttr(articleTitle);

  // Ambil deskripsi yang ada (jika ada)
  let rawMetaDesc = $('meta[name="description"], meta[property="description"]').attr('content') || '';
  let rawOgDesc = $('meta[property="og:description"]').attr('content') || '';
  let rawTwitterDesc = $('meta[name="twitter:description"]').attr('content') || '';

  // Cadangan dari paragraf pertama
  const firstP = $('p').first().text().trim();
  const fallback = firstP ? prepareDesc(firstP.substring(0, 160)) : 'Layar Kosong - Catatan dan Opini.';

  // Tentukan deskripsi terbaik (pilih yang tidak kosong, atau gunakan fallback)
  const bestMeta = rawMetaDesc || rawOgDesc || rawTwitterDesc || fallback;

  // Definisikan variabel yang dibutuhkan oleh metaTags
  const finalMetaDesc = prepareDesc(rawMetaDesc || bestMeta);
  const finalOgDesc = prepareDesc(rawOgDesc || bestMeta);
  const finalTwitterDesc = prepareDesc(rawTwitterDesc || bestMeta);



  const cleanFileName = baseName.replace('.html', '');
  const canonicalUrl = `${baseUrl}/artikel/${cleanFileName}`.replace(/\/$/, '');

  const firstP = $('p').first().text().trim();
  const fallback = firstP ? prepareDesc(firstP.substring(0, 160)) : 'Layar Kosong - Catatan dan Opini.';

  const rawMetaDesc = $('meta[name="description"]').attr('content') || '';
  const finalMetaDesc = prepareDesc(rawMetaDesc || fallback);

  let metaImgUrl = $('meta[property="og:image"]').attr('content') || $('img').first().attr('src') || '';
  if (metaImgUrl && metaImgUrl.startsWith('http')) {
    const mirroredPath = await mirrorAndConvert(metaImgUrl, baseUrl);
    if (mirroredPath.startsWith('/')) metaImgUrl = `${baseUrl}${mirroredPath}`;
  }

  // 3. BERSIHKAN & SUNTIK ULANG (Gaya Ringan)
  $('html').attr('lang', 'id').attr('prefix', 'og: https://ogp.me/ns# article: https://ogp.me/ns/article#');

  // Hapus tag lama
  $('link[rel="canonical"], link[rel="icon"], meta[name="description"], meta[property^="og:"], meta[name^="twitter:"]').remove();

  // --- 4. SUNTIK ULANG (Metode Array Join - Lebih Cepat) ---
  const metaTags = [
    `<meta property="og:locale" content="id_ID">`,
    `<meta property="og:site_name" content="Layar Kosong">`,
    `<link rel="icon" href="/favicon.ico">`,
    `<link rel="manifest" href="/site.webmanifest">`,
    `<link rel="canonical" href="${canonicalUrl}">`,
    `<meta property="og:url" content="${canonicalUrl}">`,
    `<meta property="twitter:url" content="${canonicalUrl}">`,
    `<meta property="twitter:domain" content="https://dalam.web.id">`,
    `<meta property="og:title" content="${escapedTitle}">`,
    `<meta name="twitter:title" content="${escapedTitle}">`,
    `<meta property="og:type" content="article">`,
    `<meta name="theme-color" content="#00b0ed">`,
    `<meta name="robots" content="index, follow, max-image-preview:large">`,
    `<meta name="author" content="Fakhrul Rijal">`,
    `<meta name="description" content="${finalMetaDesc}">`,
    `<meta property="og:description" content="${finalOgDesc}">`,
    `<meta name="twitter:description" content="${finalTwitterDesc}">`,
    `<link rel="license" href="https://creativecommons.org/publicdomain/zero/1.0/">`,
    `<meta name="twitter:creator" content="@responaja">`,
    `<meta name="bluesky:creator" content="@dalam.web.id">`,
    `<meta name="fediverse:creator" content="@frijal@mastodon.social">`,
    `<meta name="googlebot" content="max-image-preview:large">`,
    `<meta name="twitter:site" content="@responaja">`,
    `<meta property="article:author" content="https://facebook.com/frijal">`,
    `<meta property="article:publisher" content="https://facebook.com/frijalpage">`,
    `<meta property="fb:app_id" content="175216696195384">`
  ];

  // Tambahkan Meta Gambar jika ada
  if (metaImgUrl) {
    metaTags.push(
      `<meta itemprop="image" content="${metaImgUrl}">`,
      `<meta name="twitter:image" content="${metaImgUrl}">`,
      `<meta property="twitter:image" content="${metaImgUrl}">`,
      `<meta property="og:image" content="${metaImgUrl}">`,
      `<meta property="og:image:alt" content="${escapedTitle}">`,
      `<meta property="og:image:width" content="1200">`,
      `<meta property="og:image:height" content="675">`,
      `<meta name="twitter:card" content="summary_large_image">`
    );
  }

  // Tambahkan Tags artikel
  existingTags.forEach(tag => {
    metaTags.push(`<meta property="article:tag" content="${tag}">`);
  });

  // Tambahkan Time Metadata
  if (publishedTime) metaTags.push(`<meta property="article:published_time" content="${publishedTime}">`);
  if (modifiedTime) metaTags.push(`<meta property="article:modified_time" content="${modifiedTime}">`);

  // SUNTIK SEKALI JALAN! 🚀
  head.append('\n    ' + metaTags.join('\n    ') + '\n');

  // 4. BODY FIXES
  $('img').each((_, el) => {
    if (!$(el).attr('alt')) $(el).attr('alt', articleTitle);
  });

  // Simpan pakai Bun.write
  await Bun.write(file, $.html());
}

async function fixSEO() {
  const targetFolder = process.argv[2] || 'artikel';
  const baseUrl = 'https://dalam.web.id';

  console.log('🧼 Memulai SEO Fixer (Bun Paralel Mode)...');
  const startTime = Bun.nanoseconds();

  const glob = new Glob(`${targetFolder}/*.html`);
  const files = [];
  for await (const file of glob.scan(".")) {
    files.push(file);
  }

  // Proses semua file secara paralel (Batasi kalau ribuan file agar tidak OOM)
  await Promise.all(files.map(file => processFile(file, baseUrl)));

  const duration = (Bun.nanoseconds() - startTime) / 1e9;
  console.log(`\n✅ Selesai dalam ${duration.toFixed(2)} detik!`);
}

fixSEO().catch(err => console.error(err));

File: alat/seo-fixer.rs
use anyhow::Result;
use nipper::Document;
use std::fs;
use std::path::{Path};
use url::Url;
use walkdir::WalkDir;

const BASE_URL: &str = "https://dalam.web.id";

// 1:1 dengan escapeHtmlAttr
fn escape_html_attr(text: &str) -> String {
    if text.is_empty() { return String::new(); }
    text.replace('&', "&amp;")
    .replace('"', "&quot;")
    .replace('\'', "’")
    .replace('<', "&lt;")
    .replace('>', "&gt;")
}

// 1:1 dengan prepareDesc
fn prepare_desc(text: &str) -> String {
    let re = regex::Regex::new(r"\s+").unwrap();
    let cleaned = re.replace_all(text, " ");
    escape_html_attr(cleaned.trim())
}

// 1:1 dengan mirrorAndConvert
async fn mirror_and_convert(external_url: &str, base_url: &str) -> String {
    let url = match Url::parse(external_url) {
        Ok(u) => u,
        Err(_) => return external_url.to_string(),
    };

    let base_parsed = Url::parse(base_url).unwrap();
    let hostname = url.host_str().unwrap_or("");
    let base_hostname = base_parsed.host_str().unwrap_or("");

    if hostname == base_hostname || hostname == "localhost" || hostname == "schema.org" {
        return external_url.replace(base_url, "");
    }

    let original_path = url.path();
    let ext = Path::new(original_path).extension().and_then(|s| s.to_str()).unwrap_or("").to_lowercase();
    let is_svg = ext == "svg";
    let final_ext = if is_svg { ".svg" } else { ".webp" };

    let local_path_name = if !ext.is_empty() {
        original_path.replace(&format!(".{}", ext), final_ext)
    } else {
        format!("{}{}", original_path, final_ext)
    };

    let local_path = Path::new("img").join(hostname).join(local_path_name.trim_start_matches('/'));
    let web_path = format!("/{}", local_path.to_string_lossy().replace('\\', "/"));

    if local_path.exists() { return web_path; }

    if let Some(parent) = local_path.parent() { fs::create_dir_all(parent).ok(); }

    let client = reqwest::Client::builder()
    .user_agent("Mozilla/5.0")
    .timeout(std::time::Duration::from_secs(15))
    .build().unwrap();

    if let Ok(res) = client.get(external_url).send().await {
        if let Ok(data) = res.bytes().await {
            if is_svg {
                fs::write(&local_path, data).ok();
            } else if let Ok(img) = image::load_from_memory(&data) {
                let encoder = webp::Encoder::from_image(&img).unwrap();
                fs::write(&local_path, &*encoder.encode(85.0)).ok();
            }
            return web_path;
        }
    }
    external_url.to_string()
}

#[tokio::main]
async fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let target_folder = args.get(1).map(|s| s.as_str()).unwrap_or("artikel");
    let base_url = BASE_URL.trim_end_matches('/');

    for entry in WalkDir::new(target_folder).into_iter().filter_map(|e| e.ok()) {
        if entry.path().extension().and_then(|s| s.to_str()) != Some("html") { continue; }

        let file_path = entry.path();
        let raw_content = fs::read_to_string(file_path)?;
        let base_name = file_path.file_name().unwrap().to_string_lossy().to_string();

        println!("🔍 Memproses SEO & Meta Images: {}", base_name);
        let document = Document::from(&raw_content);

        // --- 1. MIRRORING GAMBAR DALAM BODY ---
        let images = document.select("img");
        for mut img in images.iter() {
            if let Some(src) = img.attr("src") {
                let src_str = src.to_string();
                if src_str.starts_with("http") {
                    let local = mirror_and_convert(&src_str, base_url).await;
                    if local.starts_with('/') {
                        img.set_attr("src", &format!("{}{}", base_url, local));
                    }
                }
            }
        }

        // --- 2. LOGIKA DATA SEO ---
        let title_node = document.select("title");
        let article_title = title_node.text().split(" - ").next().unwrap_or("Layar Kosong").trim().to_string();
        let escaped_title = escape_html_attr(&article_title);
        let clean_file_name = base_name.replace(".html", "");
        let canonical_url = format!("{}/artikel/{}", base_url, clean_file_name);

        let raw_meta_desc = document.select("meta[name='description'], meta[property='description']").attr("content").map(|v| v.to_string()).unwrap_or_default();
        let raw_og_desc = document.select("meta[property='og:description'], meta[name='og:description']").attr("content").map(|v| v.to_string()).unwrap_or_default();
        let raw_tw_desc = document.select("meta[property='twitter:description'], meta[name='twitter:description']").attr("content").map(|v| v.to_string()).unwrap_or_default();

        let first_p = document.select("p").first().text().trim().to_string();
        let fallback = if !first_p.is_empty() {
            let limit = first_p.chars().count().min(160);
            prepare_desc(&first_p.chars().take(limit).collect::<String>())
        } else {
            "Layar Kosong - Catatan dan Opini.".to_string()
        };

        let best_meta = if !raw_meta_desc.is_empty() { raw_meta_desc }
        else if !raw_og_desc.is_empty() { raw_og_desc }
        else if !raw_tw_desc.is_empty() { raw_tw_desc }
        else { fallback };

        let final_desc = prepare_desc(&best_meta);

        let mut meta_img = document.select("meta[name='twitter:image'], meta[property='og:image']").first().attr("content").map(|v| v.to_string())
        .unwrap_or_else(|| document.select("img").first().attr("src").map(|v| v.to_string()).unwrap_or_default());

        if meta_img.starts_with("http") {
            let mirrored = mirror_and_convert(&meta_img, base_url).await;
            if mirrored.starts_with('/') { meta_img = format!("{}{}", base_url, mirrored); }
        }

        let published_time = document.select("meta[property='article:published_time']").attr("content").map(|v| v.to_string());
        let modified_time = document.select("meta[property='article:modified_time']").attr("content").map(|v| v.to_string());

        let mut tags = Vec::new();
        document.select("meta[property='article:tag']").iter().for_each(|el| {
            if let Some(t) = el.attr("content") { tags.push(t.to_string()); }
        });

        // --- 3. BERSIHKAN TAG LAMA ---
        let mut html = document.select("html");
        html.set_attr("lang", "id");
        html.set_attr("prefix", "og: https://ogp.me/ns# article: https://ogp.me/ns/article#");

        document.select("link[rel='canonical'], link[rel='icon'], link[rel='shortcut icon'], link[rel='license']").remove();
        document.select("meta[name='description'], meta[property='description'], meta[property='og:description'], meta[name='og:description'], meta[name='twitter:description'], meta[property='twitter:description']").remove();
        document.select("meta[property^='og:'], meta[name^='twitter:'], meta[property^='twitter:'], meta[property^='article:'], meta[itemprop='image']").remove();
        document.select("meta[name='author'], meta[name='robots'], meta[name='googlebot'], meta[name='theme-color'], meta[name^='bluesky:'], meta[name^='fediverse:']").remove();

        // --- 4. SUNTIK ULANG ---
        let mut head = document.select("head");
        let mut injection = Vec::new();

        // Menggunakan string literal standar ("") bukannya raw string (r#"")
        // untuk baris yang mengandung simbol # agar tidak membingungkan compiler Rust 2021.
        injection.push("<meta property=\"og:locale\" content=\"id_ID\">".to_string());
        injection.push("<meta property=\"og:site_name\" content=\"Layar Kosong\">".to_string());
        injection.push("<link rel=\"icon\" href=\"/favicon.ico\">".to_string());
        injection.push("<link rel=\"manifest\" href=\"/site.webmanifest\">".to_string());
        injection.push(format!("<link rel=\"canonical\" href=\"{}\">", canonical_url));
        injection.push(format!("<meta property=\"og:url\" content=\"{}\">", canonical_url));
        injection.push(format!("<meta property=\"twitter:url\" content=\"{}\">", canonical_url));
        injection.push("<meta property=\"twitter:domain\" content=\"https://dalam.web.id\">".to_string());
        injection.push(format!("<meta property=\"og:title\" content=\"{}\">", escaped_title));
        injection.push(format!("<meta name=\"twitter:title\" content=\"{}\">", escaped_title));
        injection.push("<meta property=\"og:type\" content=\"article\">".to_string());

        // BARIS BERMASALAH: Sekarang kita pakai string literal biasa
        injection.push("<meta name=\"theme-color\" content=\"#00b0ed\">".to_string());

        injection.push("<meta name=\"robots\" content=\"index, follow, max-image-preview:large\">".to_string());
        injection.push("<meta name=\"author\" content=\"Fakhrul Rijal\">".to_string());
        injection.push(format!("<meta name=\"description\" content=\"{}\">", final_desc));
        injection.push(format!("<meta property=\"og:description\" content=\"{}\">", final_desc));
        injection.push(format!("<meta name=\"twitter:description\" content=\"{}\">", final_desc));
        injection.push("<link rel=\"license\" href=\"https://creativecommons.org/publicdomain/zero/1.0/\">".to_string());
        injection.push("<meta name=\"twitter:creator\" content=\"@responaja\">".to_string());
        injection.push("<meta name=\"bluesky:creator\" content=\"@dalam.web.id\">".to_string());
        injection.push("<meta name=\"fediverse:creator\" content=\"@frijal@mastodon.social\">".to_string());
        injection.push("<meta name=\"googlebot\" content=\"max-image-preview:large\">".to_string());
        injection.push("<meta name=\"twitter:site\" content=\"@responaja\">".to_string());
        injection.push("<meta property=\"article:author\" content=\"https://facebook.com/frijal\">".to_string());
        injection.push("<meta property=\"article:publisher\" content=\"https://facebook.com/frijalpage\">".to_string());
        injection.push("<meta property=\"fb:app_id\" content=\"175216696195384\">".to_string());

        if !meta_img.is_empty() {
            injection.push(format!("<meta itemprop=\"image\" content=\"{}\">", meta_img));
            injection.push(format!("<meta name=\"twitter:image\" content=\"{}\">", meta_img));
            injection.push(format!("<meta property=\"twitter:image\" content=\"{}\">", meta_img));
            injection.push(format!("<meta property=\"og:image\" content=\"{}\">", meta_img));
            injection.push(format!("<meta property=\"og:image:alt\" content=\"{}\">", escaped_title));
            injection.push("<meta property=\"og:image:width\" content=\"1200\">".to_string());
            injection.push("<meta property=\"og:image:height\" content=\"675\">".to_string());
            injection.push("<meta name=\"twitter:card\" content=\"summary_large_image\">".to_string());
        }

        for tag in tags {
            injection.push(format!("<meta property=\"article:tag\" content=\"{}\">", tag));
        }

        if let Some(t) = published_time {
            injection.push(format!("<meta property=\"article:published_time\" content=\"{}\">", t));
        }
        if let Some(t) = modified_time {
            injection.push(format!("<meta property=\"article:modified_time\" content=\"{}\">", t));
        }

        head.append_html(format!("\n    {}\n", injection.join("\n    ")));

        // --- 5. BODY FIXES ---
        let body_imgs = document.select("img");
        for mut img in body_imgs.iter() {
            if img.attr("alt").is_none() {
                img.set_attr("alt", &article_title);
            }
        }

        fs::write(file_path, document.html().to_string())?;
    }
    println!("\n✅ SEO Fixer: Selesai! Deskripsi & Gambar Sosmed aman.");
    Ok(())
}
